src/tools/idlc/src/generator.c

```diff
@@ -395,12 +396,49 @@ idlc_generate(const idl_pstate_t *pstate, const idlc_generator_config_t *config)
   memset(&generator, 0, sizeof(generator));
   generator.path = file;
 
+  if(config->base_dir) {
+    char* rel_path = NULL;
+    char* old_dir = dir;
+    // If we run into a path error we can still recover and use existing dir
+    if(idl_relative_path(config->base_dir, pstate->sources->path->name, &rel_path) == IDLC_NO_MEMORY)
+      goto err_mkpath;
+    // If root is not a parent of file path, or there was an error resolving rel path
+    // TODO maybe find a common directory between the two and build from there?
+    if(rel_path == NULL || strncmp(rel_path, "..", 2) == 0)
+      goto use_dir_path;
+    dir = NULL;
+    size_t print_len = strlen(rel_path);
+    if(sep)
+      print_len -= strlen(sep);
+    if(!(dir = idl_strndup(rel_path, print_len))){
+      dir = old_dir;
+      goto use_dir_path;
+    }
+    if(old_dir && old_dir[0] != '\0')
+      free(old_dir);
+use_dir_path:
+    if(rel_path)
+      free(rel_path);
+  }
+
   sep = dir[0] == '\0' ? "" : "/";
-  if (idl_asprintf(&generator.header.path, "%s%s%s.h", dir, sep, basename) < 0)
+  if(config->output_dir && config->output_dir[0] != '\0') {
+    if(idl_asprintf(&output_path, "%s%s%s", config->output_dir, sep, dir) < 0)
+      goto skip_output_dir;
```

Comment:
Why would it fail, other than out-of-memory? I'd bail out and avoid the somewhat inelegant `goto` in the process.